spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/msi/AzureInstanceMetadataService.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE in the project root for
+ * license information.
+ */
+package com.microsoft.azure.spring.cloud.config.msi;
+
+import org.springframework.util.Assert;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+
+/**
+ * Wraps functions provided by Azure Instance Metadata Service(IMDS)
+ */
+public class AzureInstanceMetadataService {
+    private String apiVersion = "2018-02-01-preview";
+    private String tokenEndpoint = "http://169.254.169.254/metadata/identity/oauth2/token";
+    private String objectId;
+    private String clientId;
+
+    public String buildTokenUrl(String resource) {
+        Assert.hasText(resource, "Target resource should not be empty.");
+
+        StringBuilder params = new StringBuilder();
+        try {
+            params.append(String.format("api-version=%s&resource=%s", encodeFragment(this.apiVersion),
+                    encodeFragment(resource)));
+
+            if (this.objectId != null) {
+                params.append(String.format("&object_id=%s", encodeFragment(this.objectId)));
+            }
+            if (this.clientId != null) {
+                params.append(String.format("&client_id=%s", encodeFragment(this.clientId)));
+            }
+        } catch (UnsupportedEncodingException e) {
+            throw new IllegalStateException("Failed to generate retrieve token url.", e);
+        }
+
+        return String.format(tokenEndpoint + "?%s", params.toString());
+    }
+
+    private String encodeFragment(String urlFragment) throws UnsupportedEncodingException {
+        Assert.hasText(urlFragment, "Url fragment to be encoded can not be empty.");
+        return URLEncoder.encode(urlFragment, "UTF-8");
+    }
+
+    public AzureInstanceMetadataService withApiVersion(String apiVersion) {
+        Assert.hasText(apiVersion, "API version should not be emtpy.");
+        this.apiVersion = apiVersion;
+        return this;
+    }
+
+    public AzureInstanceMetadataService withTokenEndpoint(String endpoint) {
+        Assert.hasText(endpoint, "Endpoint should not be emtpy.");
```

Comment:
Do we need to validate endpoint?

Reply:
Because the endpoint is configurable here, so did not use final.

Reply:
Yeah, I found out and deleted that comment.

Reply:
Added url validation.